src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java

```diff
@@ -0,0 +1,288 @@
+            .baseDelay(Duration.ofMillis(200L))
+            .maxBackoffTime(Duration.ofSeconds(5L))
+            .build();
+
+    final RetryCondition retryCondition;
+
+    {
+        final Set<Integer> RETRYABLE_STATUS_CODES = Stream.of(
+                HttpStatusCode.INTERNAL_SERVER_ERROR,
+                HttpStatusCode.BAD_GATEWAY,
+                HttpStatusCode.SERVICE_UNAVAILABLE,
+                HttpStatusCode.GATEWAY_TIMEOUT
+        ).collect(Collectors.toSet());
+
+        final Set<Class<? extends Exception>> RETRYABLE_EXCEPTIONS = Stream.of(
+                RetryableException.class,
+                IOException.class,
+                ApiCallAttemptTimeoutException.class,
+                ApiCallTimeoutException.class).collect(Collectors.toSet());
+
+        retryCondition = OrRetryCondition.create(
+                RetryOnStatusCodeCondition.create(RETRYABLE_STATUS_CODES),
+                RetryOnExceptionsCondition.create(RETRYABLE_EXCEPTIONS),
+                RetryOnClockSkewCondition.create(),
+                RetryOnThrottlingCondition.create()
+        );
+    }
+
+    Logger logger = LoggerFactory.getLogger("S3ClientStoreProvider");
+
+
+    /**
+     * This method returns a universal client (i.e. not bound to any region)
+     * that can be used by certain S3 operations for discovery.
+     * This is the same as universalClient(false);
+     *
+     * @return a S3Client not bound to a region
+     */
+    public S3Client universalClient() {
+        return universalClient(false);
+    }
+
+    /**
+     * This method returns a universal client (i.e.not bound to any region)
+     * that can be used by certain S3 operations for discovery
+     *
+     * @param async true to return an asynchronous client, false otherwise
+     * @return a S3Client not bound to a region
+     */
+    public <T extends AwsClient> T universalClient(boolean async) {
+        return (T)((async) ? DEFAULT_ASYNC_CLIENT : DEFAULT_CLIENT);
+    }
+
+    /**
+     * Generate a client for the named bucket using a default client to determine the location of the named bucket
+     * @param bucketName the named of the bucket to make the client for
+     * @return an S3 client appropriate for the region of the named bucket
+     */
+    protected S3Client generateClient(String bucketName){
+        //
+        // TODO: use generics like in universalClient()
+        //
+        return generateClient(bucketName, universalClient());
+    }
+
+    /**
+     * Generate an asynchronous client for the named bucket using a default client to determine the location of the named bucket
+     * @param bucketName the named of the bucket to make the client for
+     * @return an asynchronous S3 client appropriate for the region of the named bucket
+     */
+    protected S3AsyncClient generateAsyncClient(String bucketName){
+        return generateAsyncClient(bucketName, DEFAULT_CLIENT);
```

Comment:
This is coming as well from the old code in S3ClientStore. I assumed it was wanted because used primarily to locate a bucket, but let me know if we need to change it. I would do it anyway post merge to make the merge easier.